File: backend/routers/channel_groups.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.database import get_db
from ..models.server import ChannelGroup, Server, Channel
from .deps import CurrentUser, AdminUser


router = APIRouter(
    prefix="/api/servers/{server_id}/channel-groups", tags=["channel-groups"]
)


class ChannelGroupCreate(BaseModel):
    name: str


class ChannelGroupUpdate(BaseModel):
    name: str | None = None


class ChannelGroupResponse(BaseModel):
    id: int
    server_id: int
    name: str
    position: int

    class Config:
        from_attributes = True


@router.get("", response_model=list[ChannelGroupResponse])
async def list_channel_groups(
    server_id: int, user: CurrentUser, db: AsyncSession = Depends(get_db)
):
    """List all channel groups in a server."""
    result = await db.execute(
        select(ChannelGroup)
        .where(ChannelGroup.server_id == server_id)
        .order_by(ChannelGroup.position)
    )
    groups = result.scalars().all()
    return [
        ChannelGroupResponse(
            id=g.id, server_id=g.server_id, name=g.name, position=g.position
        )
        for g in groups
    ]


@router.post(
    "", response_model=ChannelGroupResponse, status_code=status.HTTP_201_CREATED
)
async def create_channel_group(
    server_id: int,
    payload: ChannelGroupCreate,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Create a new channel group (admin only).

    New groups are placed at the end of the top-level list.
    """
    # Verify server exists
    server_result = await db.execute(select(Server).where(Server.id == server_id))
    if not server_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    # Get max top-level position from both groups and ungrouped channels
    group_max = await db.execute(
        select(func.max(ChannelGroup.position)).where(
            ChannelGroup.server_id == server_id
        )
    )
    channel_max = await db.execute(
        select(func.max(Channel.top_position)).where(
            Channel.server_id == server_id, Channel.group_id == None
        )
    )
    max_pos = max(group_max.scalar() or -1, channel_max.scalar() or -1)

    group = ChannelGroup(
        server_id=server_id,
        name=payload.name,
        position=max_pos + 1,
    )
    db.add(group)
    await db.flush()

    return ChannelGroupResponse(
        id=group.id,
        server_id=group.server_id,
        name=group.name,
        position=group.position,
    )


@router.patch("/{group_id}", response_model=ChannelGroupResponse)
async def update_channel_group(
    server_id: int,
    group_id: int,
    payload: ChannelGroupUpdate,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Update channel group properties (admin only)."""
    result = await db.execute(
        select(ChannelGroup).where(
            ChannelGroup.id == group_id, ChannelGroup.server_id == server_id
        )
    )
    group = result.scalar_one_or_none()
    if not group:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Channel group not found"
        )

    if payload.name is not None:
        group.name = payload.name

    await db.flush()

    return ChannelGroupResponse(
        id=group.id, server_id=group.server_id, name=group.name, position=group.position
    )


class ReorderGroupChannelsRequest(BaseModel):
    """Request to reorder channels within a group."""

    channel_ids: list[int]


@router.post("/{group_id}/reorder-channels")
async def reorder_group_channels(
    server_id: int,
    group_id: int,
    payload: ReorderGroupChannelsRequest,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Reorder channels within a specific group.

    Only accepts channel IDs that belong to this group.
    """
    # Verify group exists
    group_result = await db.execute(
        select(ChannelGroup).where(
            ChannelGroup.id == group_id, ChannelGroup.server_id == server_id
        )
    )
    if not group_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Channel group not found"
        )

    # Fetch all channels in this group
    result = await db.execute(
        select(Channel)
        .where(Channel.server_id == server_id, Channel.group_id == group_id)
        .order_by(Channel.position)
    )
    channels = result.scalars().all()
    id_map = {c.id: c for c in channels}

    provided_ids = payload.channel_ids
    provided_set = set(provided_ids)
    existing_set = set(id_map.keys())

    # Validate: all provided IDs must be in this group
    if not provided_set.issubset(existing_set):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="channel_ids must refer to channels in this group",
        )

    # Validate: must provide all channels in the group
    if provided_set != existing_set:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="All channel IDs in the group must be provided",
        )

    # Assign new positions
    for idx, cid in enumerate(provided_ids):
        id_map[cid].position = idx

    await db.flush()

    return {"success": True}


@router.delete("/{group_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_channel_group(
    server_id: int, group_id: int, user: AdminUser, db: AsyncSession = Depends(get_db)
):
    """Delete a channel group (admin only). Channels in the group will become ungrouped."""
    result = await db.execute(
        select(ChannelGroup).where(
            ChannelGroup.id == group_id, ChannelGroup.server_id == server_id
        )
    )
    group = result.scalar_one_or_none()
    if not group:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Channel group not found"
        )

    # Get max top-level position for placing ungrouped channels
    group_max = await db.execute(
        select(func.max(ChannelGroup.position)).where(
            ChannelGroup.server_id == server_id
        )
    )
    channel_max = await db.execute(
        select(func.max(Channel.top_position)).where(
            Channel.server_id == server_id, Channel.group_id == None
        )
    )
    max_pos = max(group_max.scalar() or -1, channel_max.scalar() or -1)

    # Set all channels in this group to ungrouped and assign top_position
    channels_result = await db.execute(
        select(Channel).where(Channel.group_id == group_id).order_by(Channel.position)
    )
    channels = channels_result.scalars().all()
    for i, channel in enumerate(channels):
        channel.group_id = None
        channel.top_position = max_pos + 1 + i

    await db.delete(group)

File: backend-go/internal/handler/channel_groups.go
package handler

import (
	"database/sql"
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"

	"github.com/RMS-Server/rms-discord-go/internal/middleware"
	"github.com/RMS-Server/rms-discord-go/internal/permission"
	"github.com/RMS-Server/rms-discord-go/internal/sso"
)

// ChannelGroupHandler handles channel group CRUD endpoints.
type ChannelGroupHandler struct {
	db  *sql.DB
	sso *sso.Client
}

func NewChannelGroupHandler(db *sql.DB, sso *sso.Client) *ChannelGroupHandler {
	return &ChannelGroupHandler{db: db, sso: sso}
}

type channelGroupCreateReq struct {
	Name          string `json:"name"`
	MinLevel      int    `json:"min_level"`
	PermMinLevel  int    `json:"perm_min_level"`
	LogicOperator string `json:"logic_operator"`
}

type channelGroupUpdateReq struct {
	Name          *string `json:"name"`
	MinLevel      *int    `json:"min_level"`
	PermMinLevel  *int    `json:"perm_min_level"`
	LogicOperator *string `json:"logic_operator"`
}

type channelGroupResponse struct {
	ID            int64  `json:"id"`
	ServerID      int64  `json:"server_id"`
	Name          string `json:"name"`
	Position      int    `json:"position"`
	MinLevel      int    `json:"min_level"`
	PermMinLevel  int    `json:"perm_min_level"`
	LogicOperator string `json:"logic_operator"`
}

// ListChannelGroups returns groups filtered by user permission.
// GET /api/servers/:server_id/channel-groups
func (h *ChannelGroupHandler) ListChannelGroups(c echo.Context) error {
	user := middleware.GetUser(c)
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	rows, err := h.db.Query(
		"SELECT id, server_id, name, position, min_level, perm_min_level, logic_operator FROM channel_groups WHERE server_id = ? ORDER BY position",
		serverID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer rows.Close()

	var groups []channelGroupResponse
	for rows.Next() {
		var g channelGroupResponse
		if err := rows.Scan(&g.ID, &g.ServerID, &g.Name, &g.Position, &g.MinLevel, &g.PermMinLevel, &g.LogicOperator); err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
		}
		rule := permission.PermRule{PermMinLevel: g.PermMinLevel, GroupMinLevel: g.MinLevel, LogicOperator: g.LogicOperator}
		if permission.CanAccess(user, rule) {
			groups = append(groups, g)
		}
	}
	if groups == nil {
		groups = []channelGroupResponse{}
	}
	return c.JSON(http.StatusOK, groups)
}

// CreateChannelGroup creates a new channel group.
// POST /api/servers/:server_id/channel-groups
func (h *ChannelGroupHandler) CreateChannelGroup(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	var req channelGroupCreateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}
	if req.Name == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "name is required"})
	}

	if req.MinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_level must be >= 0"})
	}
	if req.PermMinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "perm_min_level must be >= 0"})
	}

	// Default logic operator
	if req.LogicOperator == "" {
		req.LogicOperator = "AND"
	}
	if req.LogicOperator != "AND" && req.LogicOperator != "OR" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "logic_operator must be AND or OR"})
	}

	// Verify server exists
	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM servers WHERE id = ?", serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}

	// Get max top-level position
	var groupMax, chMax sql.NullInt64
	h.db.QueryRow("SELECT MAX(position) FROM channel_groups WHERE server_id = ?", serverID).Scan(&groupMax)
	h.db.QueryRow("SELECT MAX(top_position) FROM channels WHERE server_id = ? AND group_id IS NULL", serverID).Scan(&chMax)
	gm, cm := int64(-1), int64(-1)
	if groupMax.Valid {
		gm = groupMax.Int64
	}
	if chMax.Valid {
		cm = chMax.Int64
	}
	maxPos := gm
	if cm > maxPos {
		maxPos = cm
	}
	position := int(maxPos) + 1

	res, err := h.db.Exec(
		"INSERT INTO channel_groups (server_id, name, position, min_level, perm_min_level, logic_operator) VALUES (?, ?, ?, ?, ?, ?)",
		serverID, req.Name, position, req.MinLevel, req.PermMinLevel, req.LogicOperator,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	gID, _ := res.LastInsertId()

	return c.JSON(http.StatusCreated, channelGroupResponse{
		ID: gID, ServerID: serverID, Name: req.Name,
		Position: position, MinLevel: req.MinLevel,
		PermMinLevel: req.PermMinLevel, LogicOperator: req.LogicOperator,
	})
}

// UpdateChannelGroup updates group properties.
// PATCH /api/servers/:server_id/channel-groups/:id
func (h *ChannelGroupHandler) UpdateChannelGroup(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}
	groupID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid group id"})
	}

	var g channelGroupResponse
	err = h.db.QueryRow(
		"SELECT id, server_id, name, position, min_level, perm_min_level, logic_operator FROM channel_groups WHERE id = ? AND server_id = ?",
		groupID, serverID,
	).Scan(&g.ID, &g.ServerID, &g.Name, &g.Position, &g.MinLevel, &g.PermMinLevel, &g.LogicOperator)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "channel group not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	var req channelGroupUpdateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}

	if req.Name != nil {
		g.Name = *req.Name
	}
	if req.MinLevel != nil {
		if *req.MinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_level must be >= 0"})
		}
		g.MinLevel = *req.MinLevel
	}
	if req.PermMinLevel != nil {
		if *req.PermMinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "perm_min_level must be >= 0"})
		}
		g.PermMinLevel = *req.PermMinLevel
	}
	if req.LogicOperator != nil {
		if *req.LogicOperator != "AND" && *req.LogicOperator != "OR" {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "logic_operator must be AND or OR"})
		}
		g.LogicOperator = *req.LogicOperator
	}

	_, err = h.db.Exec(
		"UPDATE channel_groups SET name = ?, min_level = ?, perm_min_level = ?, logic_operator = ? WHERE id = ?",
		g.Name, g.MinLevel, g.PermMinLevel, g.LogicOperator, groupID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, g)
}

type reorderGroupChannelsReq struct {
	ChannelIDs []int64 `json:"channel_ids"`
}

// ReorderGroupChannels reorders channels within a group.
// POST /api/servers/:server_id/channel-groups/:id/reorder-channels
func (h *ChannelGroupHandler) ReorderGroupChannels(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}
	groupID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid group id"})
	}

	// Verify group exists
	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM channel_groups WHERE id = ? AND server_id = ?", groupID, serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "channel group not found"})
	}

	var req reorderGroupChannelsReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}

	// Fetch channels in group
	rows, err := h.db.Query("SELECT id FROM channels WHERE server_id = ? AND group_id = ?", serverID, groupID)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	existingIDs := map[int64]bool{}
	for rows.Next() {
		var id int64
		rows.Scan(&id)
		existingIDs[id] = true
	}
	rows.Close()

	providedSet := map[int64]bool{}
	for _, id := range req.ChannelIDs {
		if !existingIDs[id] {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "channel_ids must refer to channels in this group"})
		}
		if providedSet[id] {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "duplicate channel id"})
		}
		providedSet[id] = true
	}
	if len(providedSet) != len(existingIDs) {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "all channel IDs in the group must be provided"})
	}

	tx, err := h.db.Begin()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer tx.Rollback()

	for idx, id := range req.ChannelIDs {
		tx.Exec("UPDATE channels SET position = ? WHERE id = ?", idx, id)
	}
	if err := tx.Commit(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, map[string]bool{"success": true})
}

// DeleteChannelGroup deletes a group, ungrouping its channels.
// DELETE /api/servers/:server_id/channel-groups/:id
func (h *ChannelGroupHandler) DeleteChannelGroup(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}
	groupID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid group id"})
	}

	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM channel_groups WHERE id = ? AND server_id = ?", groupID, serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "channel group not found"})
	}

	// Get max top-level position for ungrouping channels
	var groupMax, chMax sql.NullInt64
	h.db.QueryRow("SELECT MAX(position) FROM channel_groups WHERE server_id = ?", serverID).Scan(&groupMax)
	h.db.QueryRow("SELECT MAX(top_position) FROM channels WHERE server_id = ? AND group_id IS NULL", serverID).Scan(&chMax)
	gm, cm := int64(-1), int64(-1)
	if groupMax.Valid {
		gm = groupMax.Int64
	}
	if chMax.Valid {
		cm = chMax.Int64
	}
	maxPos := gm
	if cm > maxPos {
		maxPos = cm
	}

	tx, err := h.db.Begin()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer tx.Rollback()

	// Ungroup channels
	rows, err := tx.Query("SELECT id FROM channels WHERE group_id = ? ORDER BY position", groupID)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	var channelIDs []int64
	for rows.Next() {
		var id int64
		rows.Scan(&id)
		channelIDs = append(channelIDs, id)
	}
	rows.Close()

	for i, id := range channelIDs {
		tx.Exec("UPDATE channels SET group_id = NULL, top_position = ?, position = 0 WHERE id = ?", int(maxPos)+1+i, id)
	}

	tx.Exec("DELETE FROM channel_groups WHERE id = ?", groupID)

	if err := tx.Commit(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.NoContent(http.StatusNoContent)
}
